sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -0,0 +1,160 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.MetricNameFilter;
+import org.apache.beam.sdk.metrics.MetricQueryResults;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.metrics.MetricsFilter;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.Row;
+import org.apache.calcite.adapter.enumerable.EnumerableRel;
+import org.apache.calcite.adapter.enumerable.EnumerableRelImplementor;
+import org.apache.calcite.adapter.enumerable.PhysType;
+import org.apache.calcite.adapter.enumerable.PhysTypeImpl;
+import org.apache.calcite.linq4j.Enumerable;
+import org.apache.calcite.linq4j.Linq4j;
+import org.apache.calcite.linq4j.tree.BlockBuilder;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.plan.ConventionTraitDef;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.convert.ConverterImpl;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rel.type.RelDataType;
+
+/**
+ * BeamRelNode to replace a {@code Enumerable} node.
+ */
+public class BeamEnumerableConverter extends ConverterImpl implements EnumerableRel {
+
+  public BeamEnumerableConverter(RelOptCluster cluster, RelTraitSet traits, RelNode input) {
+    super(cluster, ConventionTraitDef.INSTANCE, traits, input);
+  }
+
+  @Override
+  public RelNode copy(RelTraitSet traitSet, List<RelNode> inputs) {
+    return new BeamEnumerableConverter(getCluster(), traitSet, sole(inputs));
+  }
+
+  @Override
+  public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    // This should always be a last resort.
+    return planner.getCostFactory().makeHugeCost();
+  }
+
+  @Override
+  public Result implement(EnumerableRelImplementor implementor, Prefer prefer) {
+    final BlockBuilder list = new BlockBuilder();
+    final RelDataType rowType = getRowType();
+    final PhysType physType =
+        PhysTypeImpl.of(implementor.getTypeFactory(), rowType, prefer.preferArray());
+    final Expression node = implementor.stash((BeamRelNode) getInput(), BeamRelNode.class);
+    list.add(Expressions.call(BeamEnumerableConverter.class, "toEnumerable", node));
+    return implementor.result(physType, list.toBlock());
+  }
+
+  public static Enumerable<Object> toEnumerable(BeamRelNode node) {
+    PipelineOptions options = PipelineOptionsFactory.create();
```

Comment:
Just checking - since `Collector` only works on the `DirectRunner` it seems fine to hardcode it here. But are the options specified elsewhere when using SQL so here it would just be validation that the configuration is supported?

Reply:
Pipelines that aren't supported should get a `NullPointerException` when the `Collector` tries to process an element. I considered adding a check but decided it could wait because this is the only public path in. Is the user able to change the runner without the ability to set the `PipelineOptions`?

Reply:
I'd rather get the options plumbed deliberately here. Assuming decent testing, it won't get left forever, but it does leave the code in a sort of weird stage. Does it add a ton of scope to plumb them?

Reply:
Plumbing the `PipelineOptions` into `BeamEnumerableConverter` is not easy as they will have to take a trip through the calcite JDBC interface on the way here (possibly on the jdbc connection string). I can add them as an argument to `toEnumerable`, but that will just move this line up to `implement`.